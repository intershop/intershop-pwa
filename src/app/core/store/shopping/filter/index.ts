// API to access ngrx filter state
export * from './filter.actions';
export * from './filter.selectors';
