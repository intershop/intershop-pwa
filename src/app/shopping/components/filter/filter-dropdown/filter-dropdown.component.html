<div class="filter-group">

  <h3 (click)="isCollapsed = !isCollapsed" [attr.aria-expanded]="!isCollapsed" aria-controls="collapseFilter">
    {{filterElement.name}}
    <!-- <span class="glyphicon float-right" [ngClass]="{'glyphicon-chevron-down' : !isCollapsed, 'glyphicon-chevron-right' : isCollapsed}"></span> -->
    <fa-icon *ngIf="!isCollapsed; else collapsed" class="float-right" [icon]="['fas', 'angle-down']"></fa-icon>
    <ng-template #collapsed>
      <fa-icon class="float-right" [icon]="['fas', 'angle-right']"></fa-icon>
    </ng-template>
  </h3>

  <ul class="filter-list">
    <ng-container *ngFor="let facet of filterElement.facets">
      <li *ngIf="facet.selected && facet.name !== 'Show all'" class="filter-item filter-selected">
        <span class="icon-selected"></span>{{facet.link.title}}
        <span class="filter-count">({{facet.count}})</span>
        <!--- TODO: implement delete filter -->
        <a class="filter-clear">
          <!-- <span class="glyphicon glyphicon-minus float-right"></span> -->
          <fa-icon class="float-right" [icon]="['fas', 'minus']"></fa-icon>
        </a>
      </li>
    </ng-container>
  </ul>

  <div id="collapseFilter" [collapse]="isCollapsed" [attr.data-testing-id]="'collapseFilter' + filterElement.name">
    <form role="filter" name="Filter" [formGroup]="filterForm">
      <select formControlName="filterDropdown" name="SearchParameter" class="ws-submit-search-param form-control">
        <option value="">{{ 'search.filter.select_option.text' | translate }}</option>
        <ng-container *ngFor="let facet of filterElement.facets">
          <option *ngIf="!facet.selected  && facet.name !== 'Show all'" [value]="facet.name">{{facet.link.title}}</option>
        </ng-container>
      </select>
    </form>
  </div>

</div>
