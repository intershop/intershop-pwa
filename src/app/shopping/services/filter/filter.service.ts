import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiService, unpackEnvelope } from '../../../core/services/api/api.service';
import { Category } from '../../../models/category/category.model';
import { FilterNavigationData } from '../../../models/filter-navigation/filter-navigation.interface';
import { FilterNavigationMapper } from '../../../models/filter-navigation/filter-navigation.mapper';
import { FilterNavigation } from '../../../models/filter-navigation/filter-navigation.model';
import { Link } from '../../../models/link/link.model';
import { SearchParameterMapper } from '../../../models/search-parameter/search-parameter.mapper';
import { SearchParameter } from '../../../models/search-parameter/search-parameter.model';

@Injectable({ providedIn: 'root' })
export class FilterService {
  constructor(private apiService: ApiService) {}

  getFilterForCategory(category: Category): Observable<FilterNavigation> {
    const idList = category.uniqueId.split('.');
    // TODO from REST
    const categoryDomainName = this.getDomainId(idList[0]);
    const params = new HttpParams()
      .set('CategoryDomainName', categoryDomainName)
      .set('CategoryName', idList[idList.length - 1]);
    return this.apiService
      .get<FilterNavigationData>('filters', { params })
      .pipe(map(filter => FilterNavigationMapper.fromData(filter)));
  }

  getFilterForSearch(searchTerm: string): Observable<FilterNavigation> {
    const searchParameter = SearchParameterMapper.toData({ queryTerm: searchTerm } as SearchParameter);
    return this.apiService
      .get<FilterNavigationData>(`filters/blablub;SearchParameter=${searchParameter}`)
      .pipe(map(filter => FilterNavigationMapper.fromData(filter)));
  }

  applyFilter(filterName: string, searchParameter: string): Observable<FilterNavigation> {
    return this.apiService
      .get<FilterNavigationData>(`filters/${filterName};SearchParameter=${searchParameter}`)
      .pipe(map(filter => FilterNavigationMapper.fromData(filter)));
  }

  getProductSkusForFilter(filterName: string, searchParameter: string): Observable<string[]> {
    return this.apiService
      .get<{ elements: Link[] }>(`filters/${filterName};SearchParameter=${searchParameter}/hits`)
      .pipe(
        unpackEnvelope(),
        map((e: Link[]) => e.map(n => n.uri.split('/')[1]))
      );
  }

  private getDomainId(rootName: string) {
    console.log(rootName);
    if (rootName === 'Specials' || rootName === 'Cameras') {
      return 'inSPIRED-inTRONICS-' + rootName;
    }
    return 'inSPIRED-' + rootName;
  }
}
