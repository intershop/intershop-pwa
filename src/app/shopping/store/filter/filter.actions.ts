import { Action } from '@ngrx/store';
import { Category } from '../../../models/category/category.model';
import { FilterNavigation } from '../../../models/filter-navigation/filter-navigation.model';
import { HttpError } from '../../../models/http-error/http-error.model';

export enum FilterActionTypes {
  LoadFilterForCategory = '[Shopping] Load Filter For Category',
  LoadFilterForCategorySuccess = '[Shopping] Load Filter For Category Success',
  LoadFilterForCategoryFail = '[Shopping] Load Filter For Category Fail',
  ApplyFilter = '[Shopping] Apply Filter',
  ApplyFilterSuccess = '[Shopping] Apply Filter Success',
  ApplyFilterFail = '[Shopping] Apply Filter Fail',
  SetFilteredProducts = '[Shopping] Set Filtered Products',
  LoadFilterForSearch = '[Shopping] Load Filter for Search',
  LoadFilterForSearchSuccess = '[Shopping] Load Filter for Search Success',
  LoadFilterForSearchFail = '[Shopping] Load Filter for Search Fail',
}

export class LoadFilterForCategory implements Action {
  readonly type = FilterActionTypes.LoadFilterForCategory;
  constructor(public payload: Category) {}
}

export class LoadFilterForCategorySuccess implements Action {
  readonly type = FilterActionTypes.LoadFilterForCategorySuccess;
  constructor(public payload: FilterNavigation) {}
}

export class LoadFilterForCategoryFail implements Action {
  readonly type = FilterActionTypes.LoadFilterForCategoryFail;
  constructor(public payload: HttpError) {}
}

export class LoadFilterForSearch implements Action {
  readonly type = FilterActionTypes.LoadFilterForSearch;
  constructor(public payload: string) {}
}

export class LoadFilterForSearchSuccess implements Action {
  readonly type = FilterActionTypes.LoadFilterForSearchSuccess;
  constructor(public payload: FilterNavigation) {}
}

export class LoadFilterForSearchFail implements Action {
  readonly type = FilterActionTypes.LoadFilterForSearchFail;
  constructor(public payload: HttpError) {}
}

export class ApplyFilter implements Action {
  readonly type = FilterActionTypes.ApplyFilter;
  constructor(public payload: { filterId: string; searchParameter: string }) {}
}

export class ApplyFilterSuccess implements Action {
  readonly type = FilterActionTypes.ApplyFilterSuccess;
  constructor(public payload: { availableFilter: FilterNavigation; filterName: string; searchParameter: string }) {}
}

export class ApplyFilterFail implements Action {
  readonly type = FilterActionTypes.ApplyFilterFail;
  constructor(public payload: HttpError) {}
}

export class SetFilteredProducts implements Action {
  readonly type = FilterActionTypes.SetFilteredProducts;
  constructor(public payload: string[]) {}
}
export type FilterActions =
  | LoadFilterForCategory
  | LoadFilterForCategorySuccess
  | LoadFilterForCategoryFail
  | ApplyFilter
  | ApplyFilterSuccess
  | ApplyFilterFail
  | SetFilteredProducts
  | LoadFilterForSearch
  | LoadFilterForSearchSuccess
  | LoadFilterForSearchFail;
