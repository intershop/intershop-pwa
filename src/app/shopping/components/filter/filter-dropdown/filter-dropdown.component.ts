import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { Filter } from '../../../../models/filter/filter.model';

/**
 * The Filter Dropdown Component displays a filter group. The facets of the filter group are presented as options of a select box.
 *
 * @example
 * <ish-filter-dropdown
 *               [filterElement]="element"
 *               (applyFilter)="applyFilter($event)"
 * </ish-filter-dropdown>
 */
@Component({
  selector: 'ish-filter-dropdown',
  templateUrl: './filter-dropdown.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FilterDropdownComponent implements OnInit {
  /**
   * The filter group.
   */
  @Input()
  filterElement: Filter;
  @Output()
  applyFilter: EventEmitter<{ filterId: string; searchParameter: string }> = new EventEmitter();
  destroy$ = new Subject();
  filterForm: FormGroup;
  isCollapsed = false;

  ngOnInit() {
    this.filterForm = new FormGroup({
      filterDropdown: new FormControl(''),
    });

    this.filterForm
      .get('filterDropdown')
      .valueChanges.pipe(takeUntil(this.destroy$))
      .subscribe(filterDropdown => this.filter(filterDropdown));
  }

  /**
   * Applies a facet of the filter group and shows the new filtered result.
   */
  filter(facetName: string) {
    const facet = this.filterElement.facets.find(f => f.name === facetName);
    this.applyFilter.emit({ filterId: facet.filterId, searchParameter: facet.searchParameter });
  }
}
