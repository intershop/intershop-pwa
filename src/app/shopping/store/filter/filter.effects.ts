import { Category } from '../../../models/category/category.model';
import { FilterService } from '../../services/filter/filter.service';
import { SelectProduct } from '../products/products.actions';
import { ApplyFilterSuccess } from './filter.actions';

import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { select, Store } from '@ngrx/store';
import { of } from 'rxjs/observable/of';
import { catchError, distinctUntilKeyChanged, filter, map, mergeMap, switchMap } from 'rxjs/operators';
import { CoreState } from '../../../core/store/core.state';
import * as fromStore from '../../store/categories';
import * as filterActions from '../filter/filter.actions';
import { SearchActionTypes, SearchProductsSuccess } from '../search';
import { ShoppingState } from '../shopping.state';

@Injectable()
export class FilterEffects {
  constructor(
    private actions$: Actions,
    private store$: Store<ShoppingState | CoreState>,
    private filterService: FilterService
  ) {}

  @Effect()
  loadAvailableFilterForCategories$ = this.actions$.pipe(
    ofType(filterActions.FilterActionTypes.LoadFilterForCategory),
    mergeMap((action: filterActions.LoadFilterForCategory) =>
      this.filterService.getFilterForCategory(action.payload).pipe(
        map(filterNavigation => new filterActions.LoadFilterForCategorySuccess(filterNavigation)),
        catchError(error => of(new filterActions.LoadFilterForCategoryFail(error)))
      )
    )
  );

  @Effect()
  loadFilterForSearch$ = this.actions$.pipe(
    ofType(filterActions.FilterActionTypes.LoadFilterForSearch),
    mergeMap((action: filterActions.LoadFilterForSearch) =>
      this.filterService.getFilterForSearch(action.payload).pipe(
        map(filterNavigation => new filterActions.LoadFilterForSearchSuccess(filterNavigation)),
        catchError(error => of(new filterActions.LoadFilterForSearchFail(error)))
      )
    )
  );

  @Effect()
  loadFilterIfCategoryWasSelected$ = this.store$.pipe(
    select(fromStore.getSelectedCategory),
    filter(category => !!category),
    distinctUntilKeyChanged('uniqueId'),
    map(
      category =>
        new filterActions.LoadFilterForCategory({
          ...category,
          children: undefined,
          hasChildren: undefined,
          pathCategories: undefined,
        } as Category)
    )
  );

  @Effect()
  loadFilterForSearchIfSearchSuccess$ = this.actions$.pipe(
    ofType(SearchActionTypes.SearchProductsSuccess),
    map((action: SearchProductsSuccess) => new filterActions.LoadFilterForSearch(action.payload.searchTerm))
  );

  @Effect()
  applyFilter$ = this.actions$.pipe(
    ofType(filterActions.FilterActionTypes.ApplyFilter),
    map((f: filterActions.ApplyFilter) => f.payload),
    mergeMap(emit =>
      this.filterService.applyFilter(emit.filterId, emit.searchParameter).pipe(
        map(
          filterNavigation =>
            new filterActions.ApplyFilterSuccess(filterNavigation, emit.filterId, emit.searchParameter)
        ),
        catchError(error => of(new filterActions.ApplyFilterFail(error)))
      )
    )
  );

  @Effect()
  loadFilteredProducts$ = this.actions$.pipe(
    ofType(filterActions.FilterActionTypes.ApplyFilterSuccess),
    switchMap((action: ApplyFilterSuccess) =>
      this.filterService.getProductSkusForFilter(action.filterName, action.searchParameter).pipe(
        mergeMap((skus: string[]) => {
          // TODO: ??
          const actions = skus.map(b => new SelectProduct(b));
          return [...actions, new filterActions.SetFilteredProducts(skus)];
        })
      )
    )
  );
}
