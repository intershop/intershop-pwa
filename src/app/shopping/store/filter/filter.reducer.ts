import { FilterNavigation } from '../../../models/filter-navigation/filter-navigation.model';
import { FilterActions, FilterActionTypes } from './filter.actions';

export interface FilterState {
  loading: boolean;
  availableFilter: FilterNavigation;
  products: string[];
}

export const initialState: FilterState = {
  loading: false,
  availableFilter: undefined,
  products: undefined,
};

export function filterReducer(state = initialState, action: FilterActions): FilterState {
  switch (action.type) {
    case FilterActionTypes.LoadFilterForCategory:
    case FilterActionTypes.LoadFilterForSearch: {
      return { ...initialState, loading: true };
    }
    case FilterActionTypes.LoadFilterForCategorySuccess:
    case FilterActionTypes.LoadFilterForSearchSuccess: {
      return {
        ...state,
        availableFilter: action.payload,
        loading: false,
      };
    }
    case FilterActionTypes.LoadFilterForCategoryFail:
    case FilterActionTypes.LoadFilterForSearchFail: {
      return {
        ...state,
        availableFilter: undefined,
        loading: false,
      };
    }
    case FilterActionTypes.ApplyFilter: {
      return {
        ...state,
        loading: true,
      };
    }
    case FilterActionTypes.ApplyFilterSuccess: {
      const { availableFilter } = action.payload;
      return {
        ...state,
        availableFilter,
        loading: false,
      };
    }

    case FilterActionTypes.SetFilteredProducts: {
      return {
        ...state,
        products: action.payload,
      };
    }
  }

  return state;
}
