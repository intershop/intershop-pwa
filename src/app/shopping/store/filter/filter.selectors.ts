import { createSelector } from '@ngrx/store';
import { getProductEntities } from '../products';
import { getShoppingState, ShoppingState } from '../shopping.state';

const getFilterState = createSelector(getShoppingState, (state: ShoppingState) => state.filter);

const getFilteredProductSkus = createSelector(getFilterState, state => state.products);

export const getAvailableFilter = createSelector(getFilterState, state => state.availableFilter);

export const getLoadingStatus = createSelector(getFilterState, state => state.loading);

export const getFilteredProducts = createSelector(
  getFilteredProductSkus,
  getProductEntities,
  (skus, products) => (skus && skus.map(sku => products[sku])) || null
);

export const getNumberOfFilteredProducts = createSelector(getFilteredProductSkus, skus => (!!skus ? skus.length : 0));
