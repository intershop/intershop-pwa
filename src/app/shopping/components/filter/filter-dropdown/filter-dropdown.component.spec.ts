import { async, ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { IconModule } from '../../../../core/icon.module';
import { Filter } from '../../../../models/filter/filter.model';
import { FilterDropdownComponent } from './filter-dropdown.component';

describe('Filter Dropdown Component', () => {
  let component: FilterDropdownComponent;
  let fixture: ComponentFixture<FilterDropdownComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [TranslateModule.forRoot(), ReactiveFormsModule, CollapseModule.forRoot(), IconModule],
      declarations: [FilterDropdownComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    const filterElement = {
      name: 'Brands',
      facets: [
        { name: 'AsusName', count: 4, link: { title: 'Asus' } },
        { name: 'LogitechName', count: 5, link: { title: 'Logitech' }, selected: true },
      ],
    } as Filter;

    fixture = TestBed.createComponent(FilterDropdownComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    component.filterElement = filterElement;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });

  it(
    'should toggle unselected filter facets when filter group header is clicked',
    fakeAsync(() => {
      fixture.detectChanges();
      const filterGroupHead = fixture.nativeElement.querySelectorAll('h3')[0];
      filterGroupHead.click();
      tick(500);
      fixture.detectChanges();

      const selectedFilterFacet = element.getElementsByClassName('filter-selected')[0];
      expect(selectedFilterFacet.textContent).toContain('Logitech');

      const hiddenFilters = element.querySelector('div [data-testing-id=collapseFilterBrands]');
      expect(hiddenFilters.getAttribute('style')).toContain('display: none;');
    })
  );
});
