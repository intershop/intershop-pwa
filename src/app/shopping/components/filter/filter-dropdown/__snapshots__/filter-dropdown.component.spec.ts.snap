// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Dropdown Component should be created 1`] = `
<div
  id="root0"
  ng-version="6.0.7"
>
  <div
    class="filter-group"
  >
    <h3>
      Brands
    </h3>
    <ul
      class="filter-list"
    >
      
      
      
      
      
      
      <li
        class="filter-item filter-selected"
      >
        <span
          class="icon-selected"
        />
        Logitech 
        <span
          class="filter-count"
        >
          (5)
        </span>
        <a
          class="filter-clear"
        >
          <span
            class="glyphicon glyphicon-minus pull-right"
          />
        </a>
      </li>
      
    </ul>
    <form
      class="ng-untouched ng-pristine ng-valid"
      name="Filter"
      ng-reflect-form="[object Object]"
      novalidate=""
      role="filter"
    >
      <select
        class="ws-submit-search-param form-control ng-untouched ng-pristine ng-valid"
        formcontrolname="filterDropdown"
        name="SearchParameter"
        ng-reflect-name="filterDropdown"
      >
        <option
          ng-reflect-value=""
          value=""
        >
          search.filter.select_option.text
        </option>
        
        
        
        <option
          ng-reflect-value="AsusName"
          value="AsusName"
        >
          Asus
        </option>
        
        
        
        
      </select>
    </form>
  </div>
</div>
`;
